### PR TITLE
feat(subsystems): extract TypeHandlerLibrary.

### PR DESCRIPTION
**Pull request by DarkWeird**
### Contains

Next part of splitting engine.
It is really not subsystem, but it used everywhere in engine and subsystems (e.g. NetworkSubsystem, ConfigurationSubsystem)

This PR prevent circular dependencies for next extractions.

Changes:
Created gradle project
Move classes
Split TypeHandlerLibrary on implementation and interface

[Bonus!] I did add `Readme.md` for TypeHandlerLibrary.
[Bonus!] Moved tests and new tests.

compability for modules:
https://github.com/Terasology/Dialogs/pull/22

[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVEQgXG4gICAgVHlwZUhhbmRsZXJMaWJyYXJ5XG4gICAgVHlwZUhhbmRsZXJGYWN0b3J5XG4gICAgVHlwZUhhbmRsZXJcbiAgICBQZXJzaXN0ZW50RGF0YVxuICAgIFBlcnNpc3RlbnREYXRhU2VyaWFsaXplclxuICAgIFxuICAgIFR5cGVIYW5kbGVyTGlicmFyeSAtLS0-IHxwcm92aWRlc3wgVHlwZUhhbmRsZXJGYWN0b3J5XG4gICAgVHlwZUhhbmRsZXJGYWN0b3J5IC0tLT4gfHByb3ZpZGVzfCBUeXBlSGFuZGxlclxuICAgIFR5cGVIYW5kbGVyICYgVHlwZUhhbmRsZXJGYWN0b3J5IC0tLT4gfHJlZ2lzdGVyfCBUeXBlSGFuZGxlckxpYnJhcnlcbiAgICBcbiAgICBUeXBlSGFuZGxlciAtLS0-IHx1c2VzIGZvciBzZXJpYWxpemF0aW9ufCBQZXJzaXN0ZW50RGF0YVNlcmlhbGl6ZXJcbiAgICBUeXBlSGFuZGxlciAtLS0-IHxyZXN1bHQgb2Ygc2VyaWFsaXphdGlvbiB8IFBlcnNpc3RlbnREYXRhXG4gICAgVHlwZUhhbmRsZXIgLS0tPiB8c291cmNlIG9mIGRlc2VyaWFsaXphdGlvbiB8IFBlcnNpc3RlbnREYXRhXG4iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCIsInRoZW1lVmFyaWFibGVzIjp7ImJhY2tncm91bmQiOiJ3aGl0ZSIsInByaW1hcnlDb2xvciI6IiNFQ0VDRkYiLCJzZWNvbmRhcnlDb2xvciI6IiNmZmZmZGUiLCJ0ZXJ0aWFyeUNvbG9yIjoiaHNsKDgwLCAxMDAlLCA5Ni4yNzQ1MDk4MDM5JSkiLCJwcmltYXJ5Qm9yZGVyQ29sb3IiOiJoc2woMjQwLCA2MCUsIDg2LjI3NDUwOTgwMzklKSIsInNlY29uZGFyeUJvcmRlckNvbG9yIjoiaHNsKDYwLCA2MCUsIDgzLjUyOTQxMTc2NDclKSIsInRlcnRpYXJ5Qm9yZGVyQ29sb3IiOiJoc2woODAsIDYwJSwgODYuMjc0NTA5ODAzOSUpIiwicHJpbWFyeVRleHRDb2xvciI6IiMxMzEzMDAiLCJzZWNvbmRhcnlUZXh0Q29sb3IiOiIjMDAwMDIxIiwidGVydGlhcnlUZXh0Q29sb3IiOiJyZ2IoOS41MDAwMDAwMDAxLCA5LjUwMDAwMDAwMDEsIDkuNTAwMDAwMDAwMSkiLCJsaW5lQ29sb3IiOiIjMzMzMzMzIiwidGV4dENvbG9yIjoiIzMzMyIsIm1haW5Ca2ciOiIjRUNFQ0ZGIiwic2Vjb25kQmtnIjoiI2ZmZmZkZSIsImJvcmRlcjEiOiIjOTM3MERCIiwiYm9yZGVyMiI6IiNhYWFhMzMiLCJhcnJvd2hlYWRDb2xvciI6IiMzMzMzMzMiLCJmb250RmFtaWx5IjoiXCJ0cmVidWNoZXQgbXNcIiwgdmVyZGFuYSwgYXJpYWwiLCJmb250U2l6ZSI6IjE2cHgiLCJsYWJlbEJhY2tncm91bmQiOiIjZThlOGU4Iiwibm9kZUJrZyI6IiNFQ0VDRkYiLCJub2RlQm9yZGVyIjoiIzkzNzBEQiIsImNsdXN0ZXJCa2ciOiIjZmZmZmRlIiwiY2x1c3RlckJvcmRlciI6IiNhYWFhMzMiLCJkZWZhdWx0TGlua0NvbG9yIjoiIzMzMzMzMyIsInRpdGxlQ29sb3IiOiIjMzMzIiwiZWRnZUxhYmVsQmFja2dyb3VuZCI6IiNlOGU4ZTgiLCJhY3RvckJvcmRlciI6ImhzbCgyNTkuNjI2MTY4MjI0MywgNTkuNzc2NTM2MzEyOCUsIDg3LjkwMTk2MDc4NDMlKSIsImFjdG9yQmtnIjoiI0VDRUNGRiIsImFjdG9yVGV4dENvbG9yIjoiYmxhY2siLCJhY3RvckxpbmVDb2xvciI6ImdyZXkiLCJzaWduYWxDb2xvciI6IiMzMzMiLCJzaWduYWxUZXh0Q29sb3IiOiIjMzMzIiwibGFiZWxCb3hCa2dDb2xvciI6IiNFQ0VDRkYiLCJsYWJlbEJveEJvcmRlckNvbG9yIjoiaHNsKDI1OS42MjYxNjgyMjQzLCA1OS43NzY1MzYzMTI4JSwgODcuOTAxOTYwNzg0MyUpIiwibGFiZWxUZXh0Q29sb3IiOiJibGFjayIsImxvb3BUZXh0Q29sb3IiOiJibGFjayIsIm5vdGVCb3JkZXJDb2xvciI6IiNhYWFhMzMiLCJub3RlQmtnQ29sb3IiOiIjZmZmNWFkIiwibm90ZVRleHRDb2xvciI6ImJsYWNrIiwiYWN0aXZhdGlvbkJvcmRlckNvbG9yIjoiIzY2NiIsImFjdGl2YXRpb25Ca2dDb2xvciI6IiNmNGY0ZjQiLCJzZXF1ZW5jZU51bWJlckNvbG9yIjoid2hpdGUiLCJzZWN0aW9uQmtnQ29sb3IiOiJyZ2JhKDEwMiwgMTAyLCAyNTUsIDAuNDkpIiwiYWx0U2VjdGlvbkJrZ0NvbG9yIjoid2hpdGUiLCJzZWN0aW9uQmtnQ29sb3IyIjoiI2ZmZjQwMCIsInRhc2tCb3JkZXJDb2xvciI6IiM1MzRmYmMiLCJ0YXNrQmtnQ29sb3IiOiIjOGE5MGRkIiwidGFza1RleHRMaWdodENvbG9yIjoid2hpdGUiLCJ0YXNrVGV4dENvbG9yIjoid2hpdGUiLCJ0YXNrVGV4dERhcmtDb2xvciI6ImJsYWNrIiwidGFza1RleHRPdXRzaWRlQ29sb3IiOiJibGFjayIsInRhc2tUZXh0Q2xpY2thYmxlQ29sb3IiOiIjMDAzMTYzIiwiYWN0aXZlVGFza0JvcmRlckNvbG9yIjoiIzUzNGZiYyIsImFjdGl2ZVRhc2tCa2dDb2xvciI6IiNiZmM3ZmYiLCJncmlkQ29sb3IiOiJsaWdodGdyZXkiLCJkb25lVGFza0JrZ0NvbG9yIjoibGlnaHRncmV5IiwiZG9uZVRhc2tCb3JkZXJDb2xvciI6ImdyZXkiLCJjcml0Qm9yZGVyQ29sb3IiOiIjZmY4ODg4IiwiY3JpdEJrZ0NvbG9yIjoicmVkIiwidG9kYXlMaW5lQ29sb3IiOiJyZWQiLCJsYWJlbENvbG9yIjoiYmxhY2siLCJlcnJvckJrZ0NvbG9yIjoiIzU1MjIyMiIsImVycm9yVGV4dENvbG9yIjoiIzU1MjIyMiIsImNsYXNzVGV4dCI6IiMxMzEzMDAiLCJmaWxsVHlwZTAiOiIjRUNFQ0ZGIiwiZmlsbFR5cGUxIjoiI2ZmZmZkZSIsImZpbGxUeXBlMiI6ImhzbCgzMDQsIDEwMCUsIDk2LjI3NDUwOTgwMzklKSIsImZpbGxUeXBlMyI6ImhzbCgxMjQsIDEwMCUsIDkzLjUyOTQxMTc2NDclKSIsImZpbGxUeXBlNCI6ImhzbCgxNzYsIDEwMCUsIDk2LjI3NDUwOTgwMzklKSIsImZpbGxUeXBlNSI6ImhzbCgtNCwgMTAwJSwgOTMuNTI5NDExNzY0NyUpIiwiZmlsbFR5cGU2IjoiaHNsKDgsIDEwMCUsIDk2LjI3NDUwOTgwMzklKSIsImZpbGxUeXBlNyI6ImhzbCgxODgsIDEwMCUsIDkzLjUyOTQxMTc2NDclKSJ9fSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVEQgXG4gICAgVHlwZUhhbmRsZXJMaWJyYXJ5XG4gICAgVHlwZUhhbmRsZXJGYWN0b3J5XG4gICAgVHlwZUhhbmRsZXJcbiAgICBQZXJzaXN0ZW50RGF0YVxuICAgIFBlcnNpc3RlbnREYXRhU2VyaWFsaXplclxuICAgIFxuICAgIFR5cGVIYW5kbGVyTGlicmFyeSAtLS0-IHxwcm92aWRlc3wgVHlwZUhhbmRsZXJGYWN0b3J5XG4gICAgVHlwZUhhbmRsZXJGYWN0b3J5IC0tLT4gfHByb3ZpZGVzfCBUeXBlSGFuZGxlclxuICAgIFR5cGVIYW5kbGVyICYgVHlwZUhhbmRsZXJGYWN0b3J5IC0tLT4gfHJlZ2lzdGVyfCBUeXBlSGFuZGxlckxpYnJhcnlcbiAgICBcbiAgICBUeXBlSGFuZGxlciAtLS0-IHx1c2VzIGZvciBzZXJpYWxpemF0aW9ufCBQZXJzaXN0ZW50RGF0YVNlcmlhbGl6ZXJcbiAgICBUeXBlSGFuZGxlciAtLS0-IHxyZXN1bHQgb2Ygc2VyaWFsaXphdGlvbiB8IFBlcnNpc3RlbnREYXRhXG4gICAgVHlwZUhhbmRsZXIgLS0tPiB8c291cmNlIG9mIGRlc2VyaWFsaXphdGlvbiB8IFBlcnNpc3RlbnREYXRhXG4iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCIsInRoZW1lVmFyaWFibGVzIjp7ImJhY2tncm91bmQiOiJ3aGl0ZSIsInByaW1hcnlDb2xvciI6IiNFQ0VDRkYiLCJzZWNvbmRhcnlDb2xvciI6IiNmZmZmZGUiLCJ0ZXJ0aWFyeUNvbG9yIjoiaHNsKDgwLCAxMDAlLCA5Ni4yNzQ1MDk4MDM5JSkiLCJwcmltYXJ5Qm9yZGVyQ29sb3IiOiJoc2woMjQwLCA2MCUsIDg2LjI3NDUwOTgwMzklKSIsInNlY29uZGFyeUJvcmRlckNvbG9yIjoiaHNsKDYwLCA2MCUsIDgzLjUyOTQxMTc2NDclKSIsInRlcnRpYXJ5Qm9yZGVyQ29sb3IiOiJoc2woODAsIDYwJSwgODYuMjc0NTA5ODAzOSUpIiwicHJpbWFyeVRleHRDb2xvciI6IiMxMzEzMDAiLCJzZWNvbmRhcnlUZXh0Q29sb3IiOiIjMDAwMDIxIiwidGVydGlhcnlUZXh0Q29sb3IiOiJyZ2IoOS41MDAwMDAwMDAxLCA5LjUwMDAwMDAwMDEsIDkuNTAwMDAwMDAwMSkiLCJsaW5lQ29sb3IiOiIjMzMzMzMzIiwidGV4dENvbG9yIjoiIzMzMyIsIm1haW5Ca2ciOiIjRUNFQ0ZGIiwic2Vjb25kQmtnIjoiI2ZmZmZkZSIsImJvcmRlcjEiOiIjOTM3MERCIiwiYm9yZGVyMiI6IiNhYWFhMzMiLCJhcnJvd2hlYWRDb2xvciI6IiMzMzMzMzMiLCJmb250RmFtaWx5IjoiXCJ0cmVidWNoZXQgbXNcIiwgdmVyZGFuYSwgYXJpYWwiLCJmb250U2l6ZSI6IjE2cHgiLCJsYWJlbEJhY2tncm91bmQiOiIjZThlOGU4Iiwibm9kZUJrZyI6IiNFQ0VDRkYiLCJub2RlQm9yZGVyIjoiIzkzNzBEQiIsImNsdXN0ZXJCa2ciOiIjZmZmZmRlIiwiY2x1c3RlckJvcmRlciI6IiNhYWFhMzMiLCJkZWZhdWx0TGlua0NvbG9yIjoiIzMzMzMzMyIsInRpdGxlQ29sb3IiOiIjMzMzIiwiZWRnZUxhYmVsQmFja2dyb3VuZCI6IiNlOGU4ZTgiLCJhY3RvckJvcmRlciI6ImhzbCgyNTkuNjI2MTY4MjI0MywgNTkuNzc2NTM2MzEyOCUsIDg3LjkwMTk2MDc4NDMlKSIsImFjdG9yQmtnIjoiI0VDRUNGRiIsImFjdG9yVGV4dENvbG9yIjoiYmxhY2siLCJhY3RvckxpbmVDb2xvciI6ImdyZXkiLCJzaWduYWxDb2xvciI6IiMzMzMiLCJzaWduYWxUZXh0Q29sb3IiOiIjMzMzIiwibGFiZWxCb3hCa2dDb2xvciI6IiNFQ0VDRkYiLCJsYWJlbEJveEJvcmRlckNvbG9yIjoiaHNsKDI1OS42MjYxNjgyMjQzLCA1OS43NzY1MzYzMTI4JSwgODcuOTAxOTYwNzg0MyUpIiwibGFiZWxUZXh0Q29sb3IiOiJibGFjayIsImxvb3BUZXh0Q29sb3IiOiJibGFjayIsIm5vdGVCb3JkZXJDb2xvciI6IiNhYWFhMzMiLCJub3RlQmtnQ29sb3IiOiIjZmZmNWFkIiwibm90ZVRleHRDb2xvciI6ImJsYWNrIiwiYWN0aXZhdGlvbkJvcmRlckNvbG9yIjoiIzY2NiIsImFjdGl2YXRpb25Ca2dDb2xvciI6IiNmNGY0ZjQiLCJzZXF1ZW5jZU51bWJlckNvbG9yIjoid2hpdGUiLCJzZWN0aW9uQmtnQ29sb3IiOiJyZ2JhKDEwMiwgMTAyLCAyNTUsIDAuNDkpIiwiYWx0U2VjdGlvbkJrZ0NvbG9yIjoid2hpdGUiLCJzZWN0aW9uQmtnQ29sb3IyIjoiI2ZmZjQwMCIsInRhc2tCb3JkZXJDb2xvciI6IiM1MzRmYmMiLCJ0YXNrQmtnQ29sb3IiOiIjOGE5MGRkIiwidGFza1RleHRMaWdodENvbG9yIjoid2hpdGUiLCJ0YXNrVGV4dENvbG9yIjoid2hpdGUiLCJ0YXNrVGV4dERhcmtDb2xvciI6ImJsYWNrIiwidGFza1RleHRPdXRzaWRlQ29sb3IiOiJibGFjayIsInRhc2tUZXh0Q2xpY2thYmxlQ29sb3IiOiIjMDAzMTYzIiwiYWN0aXZlVGFza0JvcmRlckNvbG9yIjoiIzUzNGZiYyIsImFjdGl2ZVRhc2tCa2dDb2xvciI6IiNiZmM3ZmYiLCJncmlkQ29sb3IiOiJsaWdodGdyZXkiLCJkb25lVGFza0JrZ0NvbG9yIjoibGlnaHRncmV5IiwiZG9uZVRhc2tCb3JkZXJDb2xvciI6ImdyZXkiLCJjcml0Qm9yZGVyQ29sb3IiOiIjZmY4ODg4IiwiY3JpdEJrZ0NvbG9yIjoicmVkIiwidG9kYXlMaW5lQ29sb3IiOiJyZWQiLCJsYWJlbENvbG9yIjoiYmxhY2siLCJlcnJvckJrZ0NvbG9yIjoiIzU1MjIyMiIsImVycm9yVGV4dENvbG9yIjoiIzU1MjIyMiIsImNsYXNzVGV4dCI6IiMxMzEzMDAiLCJmaWxsVHlwZTAiOiIjRUNFQ0ZGIiwiZmlsbFR5cGUxIjoiI2ZmZmZkZSIsImZpbGxUeXBlMiI6ImhzbCgzMDQsIDEwMCUsIDk2LjI3NDUwOTgwMzklKSIsImZpbGxUeXBlMyI6ImhzbCgxMjQsIDEwMCUsIDkzLjUyOTQxMTc2NDclKSIsImZpbGxUeXBlNCI6ImhzbCgxNzYsIDEwMCUsIDk2LjI3NDUwOTgwMzklKSIsImZpbGxUeXBlNSI6ImhzbCgtNCwgMTAwJSwgOTMuNTI5NDExNzY0NyUpIiwiZmlsbFR5cGU2IjoiaHNsKDgsIDEwMCUsIDk2LjI3NDUwOTgwMzklKSIsImZpbGxUeXBlNyI6ImhzbCgxODgsIDEwMCUsIDkzLjUyOTQxMTc2NDclKSJ9fSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)

### How to test

1. launch game (check TypeHandlerLibrary with Gson)
2. start game (check TypeHandlerLibrary with Gson)
3. exit to main menu. (check TypeHandlerLibrary with Gson)
4. create server (check TypeHandlerLibrary with Gson)
5. join to server (check TypeHandlerLibrary with Protobuf)

### Outstanding before merging

- [ ] Javadocs

### CODE REVIEW

File: subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/SimpleCoreHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                Arguments.of(1L, new LongTypeHandler(), new PersistedLong(1)),
+                Arguments.of(1.0F, new FloatTypeHandler(), new PersistedFloat(1)),
+                Arguments.of(1.0, new DoubleTypeHandler(), new PersistedDouble(1)),
+
+                Arguments.of("foo", new StringTypeHandler(), new PersistedString("foo")),
+                Arguments.of('f', new CharacterTypeHandler(), new PersistedString("f")));
+    }
+
+    @ParameterizedTest(name = "{1}")
+    @DisplayName("Chech simple core types serialization")
```
Comment (skaldarnar) [inline subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/SimpleCoreHandlerTest.java:42]:
```suggestion
    @DisplayName("Check simple core types serialization")
```

File: subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/SimpleCoreHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                Arguments.of(true, new BooleanTypeHandler(), new PersistedBoolean(true)),
+
+                Arguments.of(1.0, new NumberTypeHandler(), new PersistedDouble(1)),
+                Arguments.of(1, new IntTypeHandler(), new PersistedInteger(1)),
+                Arguments.of(1L, new LongTypeHandler(), new PersistedLong(1)),
+                Arguments.of(1.0F, new FloatTypeHandler(), new PersistedFloat(1)),
+                Arguments.of(1.0, new DoubleTypeHandler(), new PersistedDouble(1)),
+
+                Arguments.of("foo", new StringTypeHandler(), new PersistedString("foo")),
+                Arguments.of('f', new CharacterTypeHandler(), new PersistedString("f")));
```
Comment (skaldarnar) [inline subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/SimpleCoreHandlerTest.java:38]: Testing some edge cases would be nice, but the setup is quite nice and we can easily add them later.

File: subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/inMemory/arrays/PersistedDoubleArray.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Override
+    public TFloatList getAsFloatArray() {
+        return TroveUtils.doubleToFloat(data);
+    }
+
+    @Override
+    public TIntList getAsIntegerArray() {
+        return TroveUtils.doubleToInt(data);
```
[the remark this answers is not in the document]
Comment (DarkWeird) [reply subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/inMemory/arrays/PersistedDoubleArray.java:40]: just copied behavior from protobuf and/or gson `PersistentData` impls.

Really idkm why?

File: engine/build.gradle [hunk 4]
```diff
@@ -179,11 +181,13 @@ jar {
 
     from(tasks.getByName("cacheReflections").outputs)
 
-    manifest {
-        def manifestClasspath = "$subDirLibs/" + configurations."${sourceSets.main.runtimeClasspathConfigurationName}".collect {
-            it.getName()
-        }.join(" $subDirLibs/")
-        attributes("Class-Path": manifestClasspath, "Implementation-Title": "Terasology-" + project.name, "Implementation-Version": env.BUILD_NUMBER + ", " + convertGitBranch(env.GIT_BRANCH) + ", " + env.BUILD_ID + ", " + displayVersion)
+    doFirst {
```
[the remark this answers is not in the document]
Comment (DarkWeird) [reply engine/build.gradle:184]: trying to collect RuntimeClasspath at setup gradle script is bad practics ( it happens every time when you press gradle's refresh button)
Also i catched race condition over gradle's cache  when extracted TypeHandlerLibrary.

Then I did use `doFirst` for avoid race condition and made collecting RuntimeClasspath task-related.
`doFirst`, not `doLast` - because manifest required at task's running.

 
{`@see` https://docs.gradle.org/current/userguide/tutorial_using_tasks.html}

File: subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/SimpleCoreHandlerTest.java (hunk 2 above)
Comment (DarkWeird) [reply subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/SimpleCoreHandlerTest.java:38]: this typehandlers almost haven't edge cases. (related to java's data types)

yeah, I known one case :D sends emoji character to CharacterTypeHandler.. and deserialize it back.

**Review by skaldarnar — APPROVED** [review]
Tested locally in both single and multi-player, seems to work fine!

**Comment by DarkWeird** [comment]
> Tested locally in both single and multi-player, seems to work fine!

Are we wait another review?

**Comment by skaldarnar** [comment]
Can the following error be related to this PR? Saw it when creating a new game of Metal Renegades which uses Dialogs (I've got https://github.com/Terasology/Dialogs/pull/22 checked out, too).

```
21:24:45.341 [Thread-12] ERROR o.t.module.sandbox.ModuleClassLoader - Denied access to class (not allowed with this module's permissions): org.terasology.persistence.typeHandling.TypeHandler
21:24:45.348 [Thread-12] ERROR o.t.module.sandbox.ModuleClassLoader - Denied access to class (not allowed with this module's permissions): org.terasology.persistence.typeHandling.TypeHandlerFactory
Exception in thread "Thread-12" java.lang.NullPointerException
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyHandler.deserialize(BlockFamilyDefinitionFormat.java:331)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyHandler.deserialize(BlockFamilyDefinitionFormat.java:323)
	at com.google.gson.TreeTypeAdapter.read(TreeTypeAdapter.java:58)
	at com.google.gson.Gson.fromJson(Gson.java:879)
	at com.google.gson.Gson.fromJson(Gson.java:944)
	at com.google.gson.Gson$1.deserialize(Gson.java:138)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.setObject(BlockFamilyDefinitionFormat.java:299)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.deserialize(BlockFamilyDefinitionFormat.java:142)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.deserialize(BlockFamilyDefinitionFormat.java:128)
	at com.google.gson.TreeTypeAdapter.read(TreeTypeAdapter.java:58)
	at com.google.gson.Gson.fromJson(Gson.java:879)
	at com.google.gson.Gson.fromJson(Gson.java:817)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat.load(BlockFamilyDefinitionFormat.java:93)
	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat.load(BlockFamilyDefinitionFormat.java:70)
	at org.terasology.assets.module.UnloadedAssetData$AssetSourceResolver.load(UnloadedAssetData.java:292)
	at org.terasology.assets.module.UnloadedAssetData.load(UnloadedAssetData.java:180)
	at org.terasology.assets.module.ModuleAssetDataProducer.getAssetData(ModuleAssetDataProducer.java:228)
	at org.terasology.assets.AssetType.reloadFromProducers(AssetType.java:501)
	at org.terasology.assets.AssetType.refresh(AssetType.java:182)
	at org.terasology.assets.module.ModuleAwareAssetTypeManager.switchEnvironment(ModuleAwareAssetTypeManager.java:394)
	at org.terasology.engine.bootstrap.EnvironmentSwitchHandler.handleSwitchToGameEnvironment(EnvironmentSwitchHandler.java:117)
	at org.terasology.engine.modes.loadProcesses.RegisterMods.lambda$step$0(RegisterMods.java:90)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by DarkWeird** [comment]
> Can the following error be related to this PR? Saw it when creating a new game of Metal Renegades which uses Dialogs (I've got [Terasology/Dialogs#22](https://github.com/Terasology/Dialogs/pull/22) checked out, too).
> 
> ```
> 21:24:45.341 [Thread-12] ERROR o.t.module.sandbox.ModuleClassLoader - Denied access to class (not allowed with this module's permissions): org.terasology.persistence.typeHandling.TypeHandler
> 21:24:45.348 [Thread-12] ERROR o.t.module.sandbox.ModuleClassLoader - Denied access to class (not allowed with this module's permissions): org.terasology.persistence.typeHandling.TypeHandlerFactory
> Exception in thread "Thread-12" java.lang.NullPointerException
> 	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyHandler.deserialize(BlockFamilyDefinitionFormat.java:331)
> 	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyHandler.deserialize(BlockFamilyDefinitionFormat.java:323)
> 	at com.google.gson.TreeTypeAdapter.read(TreeTypeAdapter.java:58)
> 	at com.google.gson.Gson.fromJson(Gson.java:879)
> 	at com.google.gson.Gson.fromJson(Gson.java:944)
> 	at com.google.gson.Gson$1.deserialize(Gson.java:138)
> 	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.setObject(BlockFamilyDefinitionFormat.java:299)
> 	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.deserialize(BlockFamilyDefinitionFormat.java:142)
> 	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat$BlockFamilyDefinitionDataHandler.deserialize(BlockFamilyDefinitionFormat.java:128)
> 	at com.google.gson.TreeTypeAdapter.read(TreeTypeAdapter.java:58)
> 	at com.google.gson.Gson.fromJson(Gson.java:879)
> 	at com.google.gson.Gson.fromJson(Gson.java:817)
> 	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat.load(BlockFamilyDefinitionFormat.java:93)
> 	at org.terasology.world.block.loader.BlockFamilyDefinitionFormat.load(BlockFamilyDefinitionFormat.java:70)
> 	at org.terasology.assets.module.UnloadedAssetData$AssetSourceResolver.load(UnloadedAssetData.java:292)
> 	at org.terasology.assets.module.UnloadedAssetData.load(UnloadedAssetData.java:180)
> 	at org.terasology.assets.module.ModuleAssetDataProducer.getAssetData(ModuleAssetDataProducer.java:228)
> 	at org.terasology.assets.AssetType.reloadFromProducers(AssetType.java:501)
> 	at org.terasology.assets.AssetType.refresh(AssetType.java:182)
> 	at org.terasology.assets.module.ModuleAwareAssetTypeManager.switchEnvironment(ModuleAwareAssetTypeManager.java:394)
> 	at org.terasology.engine.bootstrap.EnvironmentSwitchHandler.handleSwitchToGameEnvironment(EnvironmentSwitchHandler.java:117)
> 	at org.terasology.engine.modes.loadProcesses.RegisterMods.lambda$step$0(RegisterMods.java:90)
> 	at java.base/java.lang.Thread.run(Thread.java:834)
> ```


hehe.
Did you run `gradlew jar`? :D
there "awesome" reflections library used. it gradle level only (T_T)

**Review by skaldarnar — APPROVED** [review]
I _did_ run `gradlew jar` some time later, and wasn't able to reproduce this issue after that. So I guess this was the problem 👍
